src/metrics/function_error.py

```diff
@@ -0,0 +1,358 @@
+                x_test[j]=x_batch[j] @test_space[j]
+
+            for j in range(1, sequence_length):
+                
+                cur_x=x_context.clone()
+                cur_x[:,j]=x_test[:, j]
+
+                if isinstance(params, list):
+                    y_test=self.fn_cls.evaluate(cur_x, *params)
+                else:
+                    y_test=self.fn_cls.evaluate(cur_x, params)
+                
+                with torch.no_grad():
+                    errs[:, i, :, j] = torch.stack([
+                  
+                        self._metric(
+                            y_test,
+                            model.forward(cur_x, y_test)
+                        )
+                        for model in models
+                    ])[:, :, j]
+        
+        errs=torch.reshape(errs, (len(models), self.num_batches*batch_size, sequence_length))[:, :, 1:]
+
+        return self.PostProcessingStats(errs, [model.name for model in models]), errs
+
+        
+    def evaluateAtSeenPoints(self, models: Iterable[ContextModel]):#each evaluation happens at a random already seen point. 
+        
+        sequence_length=self.fn_cls.sequence_length
+        batch_size=self.fn_cls.batch_size
+        errs=torch.zeros((len(models), self.num_batches, batch_size, sequence_length))
+
+        for i in range(self.num_batches):
+            params=self.fn_cls.p_dist.sample()
+            x_batch=self.fn_cls.x_dist.sample()
+            
+            for j in range(1, sequence_length):
+                x_test=x_batch.clone()
+                perm = torch.stack([torch.randperm(j) for _ in range(batch_size)]).unsqueeze(dim=1)
+                ind_mat = (perm == 0) + 0.0
+                x_test[:, j:j+1] = ind_mat @ x_batch[:, :j]
+
+                if isinstance(params, list):
+                    y_test=self.fn_cls.evaluate(x_test, *params)
+                else:
+                    y_test=self.fn_cls.evaluate(x_test, params)
+                
+                with torch.no_grad():
+                    errs[:, i, :, j] = torch.stack([
+                  
+                        self._metric(
+                            y_test,
+                            model.forward(x_test, y_test)
+                        )
+                        for model in models
+                    ])[:, :, j]
+        
+        errs=torch.reshape(errs, (len(models), self.num_batches*batch_size, sequence_length))[:, :, 1:]
+
+        return self.PostProcessingStats(errs, [model.name for model in models]), errs
+
+class SquaredError(FunctionClassError):
+
+    def _metric(self, ground_truth: Tensor, predictions: Tensor) -> Tensor:
+        return (ground_truth - predictions).square()
+
+class MeanSquaredError(FunctionClassError): #i dont really think we need this, I really mostly want the local ones...
+
+    def _metric(self, ground_truth: Tensor, predictions: Tensor) -> Tensor:
+        return (ground_truth - predictions).square().mean()
```

Comment:
Discussion thread on the utility of this class

Reply:
I do not think this is necessary as it averages everything, and the Squared error is therefore just a more general version of this, without really having any drawbacks

Reply:
It also stands in contrast to the other methods which work locally

Reply:
I think the concept of averaging errors might be useful, perhaps we need to add a way to specify along which dimensions to average? E.g. what if we average errors across x dimensions?

Reply:
Yeah, maybe if one just wants a global mean for some algorithmic thing it might be useful to have this option to reduce memory usage, so I guess maybe its nice to have. But i feel like it should be general in that case, and not just for squared error

Reply:
We'll keep it for now since it *shouldn't* cause any confusion and only costs 3 lines of code.